sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,309 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Implementation of a liquidity Pool for Sui.
+///
+/// This solution is rather simple and is based on the example from the Move repo:
+/// https://github.com/move-language/move/blob/main/language/documentation/examples/experimental/coin-swap/sources/CoinSwap.move
+module defi::pool {
+    use sui::Coin::{Self, Coin, TreasuryCap};
+    use sui::Balance::{Self, Balance};
+    use sui::ID::{VersionedID};
+    use sui::SUI::SUI;
+    use sui::Transfer;
+    use sui::TxContext::{Self, TxContext};
+
+    /// The Capability to create new pools.
+    struct PoolCreatorCap has key, store {
+        id: VersionedID
+    }
+
+    /// The Pool token that will be used to mark the pool share
+    /// of a liquidity provider.
+    struct LSP<phantom T> has drop {}
+
+    /// The pool with exchange.
+    struct Pool<phantom T> has key {
+        id: VersionedID,
+        sui: Balance<SUI>,
+        token: Balance<T>,
+        lsp_treasury: TreasuryCap<LSP<T>>
+    }
+
+    /// On init creator of the module gets the capability
+    /// to create new `Pool`s.
+    fun init(ctx: &mut TxContext) {
+        Transfer::transfer(PoolCreatorCap {
+            id: TxContext::new_id(ctx)
+        }, TxContext::sender(ctx))
+    }
+
+    /// Create new `Pool` for token `T`. Each Pool holds a `Coin<T>`
+    /// and a `Coin<SUI>`. Swaps are available in both directions.
+    ///
+    /// - `share` argument defines the initial amount of LSP tokens
+    /// received by the creator of the Pool.
+    public fun create_pool<T>(
+        _: &PoolCreatorCap,
+        token: Coin<T>,
+        sui: Coin<SUI>,
+        share: u64,
+        ctx: &mut TxContext
+    ) {
+        let lsp_treasury = Coin::create_currency(LSP<T> {}, ctx);
+        let lsp = Coin::mint(share, &mut lsp_treasury, ctx);
+
+        Transfer::transfer(lsp, TxContext::sender(ctx));
+        Transfer::share_object(Pool {
+            id: TxContext::new_id(ctx),
+            token: Coin::into_balance(token),
+            sui: Coin::into_balance(sui),
+            lsp_treasury,
+        });
+    }
+
+    /// Swap `Coin<SUI>` for the `Coin<T>`.
+    public(script) fun swap_sui<T>(
+        pool: &mut Pool<T>, sui: Coin<SUI>, ctx: &mut TxContext
+    ) {
+        let sui_balance = Coin::into_balance(sui);
+
+        // Calculate the output amount - fee
+        let (sui_reserve, token_reserve, _) = get_amounts(pool);
+        let output_amount = get_input_price(
+            Balance::value(&sui_balance),
+            sui_reserve,
+            token_reserve
+        );
+
+        Balance::join(&mut pool.sui, sui_balance);
+        Transfer::transfer(
```

Comment:
This kind of thing seems ubiquitous enough that I'm wondering if it might make sense to add `Coin::withdraw_to_sender` and `Coin::mint_to_sender` (or similar) to the stdlib...